Blog post translation (You probably don't need derived state) (#101)
reactjs/fr.reactjs.org: content/blog/2018-06-07-you-probably-dont-need-derived-state.md

* [Les solution recommandées](#preferred-solutions)
* [Et la mémoïsation dans tout ça ?](#what-about-memoization)

## Quand utiliser les états dérivés {#when-to-use-derived-state}

`getDerivedStateFromProps` n'existe que dans un seul but. Il permet à un composant de mettre à jour son état interne suite à **un changement de ses props**. Notre article de blog précédent fournissait quelques examples tel que [l'enregistrement de la direction du défilement basé sur le changement d'une prop `offset`](/blog/2018/03/27/update-on-async-rendering.html#updating-state-based-on-props) ou [le chargement de données externes spécifié avec une prop source](/blog/2018/03/27/update-on-async-rendering.html#fetching-external-data-when-props-change).

Nous n'avons pas données beaucoup d'examples car, de manière général, **les états dérivés devraient être utilisés avec parcimonie**. Tous les problèmes que nous avons identifiés avec les états dérivés peuvent, au bout du bout, se résumer à (1) la mise à jour inconditionnel de l'états avec les props ou (2) la mise à jour de l'état à chaque fois que l'état et les props ne correspondent pas. (Nous allons détailler ces deux cas ci-après)

> * Si vous utilisez les états dérivés pour mémoïser certains calcules sur la seul base des props courantes vous n'avez pas besoin des états dérivés. Allez voir [Et la mémoïsation dans tout ça ƒ?](#what-about-memoization) ci-après.

Author: Oui, je comprend, mais tu sais, à ce niveau là, c'est les goûts et les couleurs, hein ;)

## Les solution recommandées {#preferred-solutions}

### Recommandation: Un composant complètement controlé {#recommendation-fully-controlled-component}

Une façon d'éviter les problèmes que nous venons de voir consiste à supprimer complètement l'état de notre composant. Si l'adresse e-mail n'existe que sous la forme d'une prop nous n'avons pas à nous soucier d'un éventuel conflit avec l'état. On peut même convertir `EmailInput` en un petit composant fonction :
```js
function EmailInput(props) {
  return <input onChange={props.onChange} value={props.email} />;
}
```

Cette approche simplifie la mise en œuvre de notre composant mais si on veut toujours stocker une valeur intermédiaire, c'est le composant de formulaire parent qui va devoir gérer ça lui-même. ([Cliquez ici pour voir la démo de cette solution.](https://codesandbox.io/s/7154w1l551))

### Recommendation: Un composant complètement non-controlé avec une `key` {#recommendation-fully-uncontrolled-component-with-a-key}

> Une autre possibilité serait que notre composant gère lui-même les états « intermédiaires » de l'e-mail. Dans ce cas notre composant pourrait toujours accepter une prop pour définir la valeur _initiale_ mais elle serait ignorée lors de tous ses changement subséquent :

Reviewer: ```suggestion
Une autre possibilité serait que notre composant gère lui-même le « brouillon » d’état de l'e-mail. Dans ce cas notre composant pourrait toujours accepter une prop pour définir la valeur _initiale_ mais tout changement ultérieur de cette prop serait ignoré :
```

}
```

> Pour pouvoir réinitialiser la valeur lorsqu'on change de contexte (comme dans notre scénario de gestionnaire de mot de passe), on peut utiliser l'attribut spécial de React appelé `key`. Quand une `key` change, React [créera une nouvelle instance du composant plutôt que de mettre à jour le composant courant](/docs/reconciliation.html#keys). Les clés sont habituellement utilisées pour les listes dynamiques mais peuvent êtres également utiles dans ce cas. Dans notre example on pourrai utiliser l'ID de l'utilisateur pour recréer le champs e-mail à chaque fois qu'un nouvel utilisateur est sélectionné :

Reviewer: ```suggestion
Pour pouvoir réinitialiser la valeur lorsqu'on change de contexte (comme dans notre scénario de gestionnaire de mots de passe), on peut utiliser l'attribut spécial de React appelé `key`. Quand une `key` change, React [créera une nouvelle instance du composant plutôt que de mettre à jour le composant actuel](/docs/reconciliation.html#keys). Les clés sont habituellement utilisées pour les listes dynamiques mais sont aussi utiles dans ce cas. Dans notre exemple on pourrait utiliser l'ID de l'utilisateur pour recréer le champ e-mail à chaque fois qu'un nouvel utilisateur est sélectionné :
```

/>
```

> A chaque fois que l'ID change, le composant `EmailInput` sera recréé et son état sera initialisé avec la dernière valeur fournis par la prop `defaultEmail`. ([Cliquez ici pour voir la démo de cette solution.](https://codesandbox.io/s/6v1znlxyxn)) Avec cette méthode, vous n'avez pas à ajouter une `key` à tous les champs. C'est d'ailleurs sans doute plus sensé de rajouter une `key` sur le formulaire lui-même plutôt que sur un simple champs. De cette manière, chaque fois que la clé changera, tous les composants dans le formulaire seront recréés avec un nouvelle état fraichement initialisé.

Reviewer: ```suggestion
À chaque fois que l'ID change, le composant `EmailInput` sera recréé et son état sera initialisé avec la dernière valeur fournie par la prop `defaultEmail`. ([Cliquez ici pour voir la démo de cette solution.](https://codesandbox.io/s/6v1znlxyxn)) Avec cette approche, vous n'avez pas à ajouter une `key` à tous les champs. C'est d'ailleurs sans doute plus sensé d’ajouter une `key` sur le formulaire lui-même plutôt que sur un simple champ. De cette manière, chaque fois que la clé changera, tous les composants dans le formulaire seront recréés avec un nouvel état fraîchement initialisé.
```

> Dans la plupart des cas, c'est la meilleur façon de gérer les états qui ont besoin d'êtres réinitialisés.

Reviewer: ```suggestion
Dans la plupart des cas, c'est la meilleure façon de gérer les états qui ont besoin d'être réinitialisés.
```

> > Remarque :

Reviewer: ```suggestion
> Remarque
```

>

> > Bien que ça semble plus lent, la différence de performance est généralement insignifiante. Utiliser une clé peut même être plus rapide si les composants on une logique lourde qui s'exécute à chaque mise à jour, là où aucun calcul d'écart n'est réalisé sur l'arbre enfant.

Reviewer: ```suggestion
> Bien que ça semble plus lent, la différence de performance est généralement insignifiante. Utiliser une clé peut même être plus rapide si les composants exécutent une logique coûteuse à chaque mise à jour, puisqu’ici la comparaison des arbres enfants est carrément sautée.
```

> #### Alternative 1: Réinitialisé un composant non-controlé avec une prop ID {#alternative-1-reset-uncontrolled-component-with-an-id-prop}

Reviewer: ```suggestion
#### Alternative 1 : réinitialiser un composant non-contrôlé avec une prop ID {#alternative-1-reset-uncontrolled-component-with-an-id-prop}
```

> Si, pour quelque raison que ce soit, `key` ne répond pas aux besoins (peut-être que le composant est très couteux à initialiser), une solution possible mais lourde serait d'observer les changements de `userID` dans `getDerivedStateFromProps` :

Reviewer: ```suggestion
Si, pour quelque raison que ce soit, `key` ne répond pas aux besoins (peut-être que le composant est très coûteux à initialiser), une solution possible mais pataude consisterait à observer les changements de `userID` dans `getDerivedStateFromProps` :
```

  };

  static getDerivedStateFromProps(props, state) {

>     // A chaque fois que l'utilisateur courant change, on réinitialise

Reviewer: ```suggestion
    // À chaque fois que l‘utilisateur actuel change, on réinitialise
```

>     // tous les aspects de l'état qui sont liés à cet utilisateur.

Reviewer: ```suggestion
    // tous les aspects de l’état qui sont liés à cet utilisateur.
```

>     // Dans cette example simple, il ne s'agit que de l'e-mail.

Reviewer: ```suggestion
    // Dans cet exemple simple, il ne s’agit que de l’e-mail.
```

}
```

> Ça donne aussi une certaine flexibilité pour ne réinitialiser qu'une partie de l'état de notre composant si c'est ce que l'on veut. ([Cliquez ici pour voir la démo de cette solution.](https://codesandbox.io/s/rjyvp7l3rq))

Reviewer: ```suggestion
Ça donne aussi une certaine flexibilité pour ne réinitialiser qu'une partie de l'état de notre composant si c'est ce que l'on veut. ([Cliquez ici pour voir une démo de cette solution.](https://codesandbox.io/s/rjyvp7l3rq))
```

> > Remarque :

Reviewer: ```suggestion
> Remarque
```

>

> > Bien que cet example n'utilise que `getDerivedStateFromProps`, la même technique peut être utilisée avec `componentWillReceiveProps`.

Reviewer: ```suggestion
> Bien que cet exemple n'utilise que `getDerivedStateFromProps`, la même technique peut être utilisée avec `componentWillReceiveProps`.
```

> #### Alternative 2: Réinitialiser un composant non-controlé avec une méthode d'instance {#alternative-2-reset-uncontrolled-component-with-an-instance-method}

Reviewer: ```suggestion
#### Alternative 2 : réinitialiser un composant non-contrôlé avec une méthode d'instance {#alternative-2-reset-uncontrolled-component-with-an-instance-method}
```

> Plus rarement vous pouvez avoir besoin de réinitialiser un état même s'il n'y a pas d'ID utilisable avec `key`. Une solution possible consiste à utiliser une clé avec une valeur aléatoire ou un nombre incrémenté automatiquement chaque fois qu'on veux réinitialiser l'état. Une autre solution consiste à exposer une méthode d'instance qui forcera la réinitialisation de l'état interne :

Reviewer: ```suggestion
Plus rarement vous pouvez avoir besoin de réinitialiser un état même s'il n'y a pas d'ID utilisable pour `key`. Une solution possible consiste à utiliser une clé avec une valeur aléatoire ou un nombre incrémenté automatiquement chaque fois qu'on veut réinitialiser l'état. Une autre solution consiste à exposer une méthode d'instance qui forcera la réinitialisation de l'état interne :
```

}
```

> Le composant de formulaire parent pourrait alors [utiliser une `ref` pour appeler cette méthode](/docs/glossary.html#refs). ([Cliquez ici pour voir la démo de cette solution.](https://codesandbox.io/s/l70krvpykl))

Reviewer: ```suggestion
Le composant de formulaire parent pourrait alors [utiliser une `ref` pour appeler cette méthode](/docs/glossary.html#refs). ([Cliquez ici pour voir une démo de cette solution.](https://codesandbox.io/s/l70krvpykl))
```

> Les références peuvent être utiles dans certains cas comme celui-ci mais, généralement, nous recommandons de les utiliser avec parcimonie. Même dans le cadre de cette démo, cette méthode impérative est sous-optimale car elle entraine la réalisation de deux rendus au lieu d'un seul.

Reviewer: ```suggestion
Les refs peuvent être utiles dans certains cas comme celui-ci mais de façon générale, nous recommandons de les utiliser avec parcimonie. Même dans le cadre de cette démo, cette méthode impérative est sous-optimale car elle entraîne la réalisation de deux rendus au lieu d'un seul.
```

-----

> ### Récapitulons {#recap}

Reviewer: ```suggestion
### En résumé {#recap}
```

Trad normalisée

> Quand on définie un composant, il est important de décider si ses données seront controlées ou non-controlées.

Reviewer: ```suggestion
Quand on définit un composant, il est important de décider si ses données seront contrôlées ou non-contrôlées.
```

> Plutôt que d'essayer de **refléter les valeur des props dans l'état**, faite un composant controlé et réconciliez les valeurs divergentes dans l'état d'un composant parent. Par example, plutôt que d'avoir un composant enfant qui accepterai une valeur « imperative » `props.value` et de suivre les valeurs intermédiaire via `state.value`, faites en sorte que le composant parent gère à la fois `state.draftValue` et `state.committedValue` pour qu'il contrôle ensuite directement la valeur du composant enfant. Ça rend le flux de données plus explicite et prédictif.

Reviewer: ```suggestion
Plutôt que d'essayer de **refléter une valeur de prop dans l'état**, faites un composant **contrôlé** et réconciliez les valeurs divergentes dans l'état d'un composant parent. Par exemple, plutôt que d'avoir un composant enfant qui accepterait une valeur « finalisée » `props.value` et garderait trace de sa valeur « brouillon » dans `state.value`, faites en sorte que le composant parent gère à la fois `state.draftValue` et `state.committedValue` pour qu'il contrôle ensuite directement la valeur du composant enfant. Ça rend le flux de données plus explicite et prévisible.
```

> Pour les composants non-controlés, si vous essayez de réinitialiser l'état quand une prop particulière (en général un ID) change, vous avez peu d'options :

Reviewer: ```suggestion
Pour les composants **non-contrôlés**, si vous essayez de réinitialiser l'état quand une prop particulière (en général un ID) change, vous avez quelques options :
```

> * **Recommandation: Réinitialiser _l'intégralité de l'état interne_ en utilisant un attribut `key`.**

Reviewer: ```suggestion
* **Recommandation : réinitialiser _l'intégralité de l'état interne_ en utilisant un attribut `key`.**
```

> * Alternative 1: Réinitialiser _certain champs de l'état_ en observant les changement d'une propriété donnée (i.e. `props.userID`).

Reviewer: * Alternative 1: To reset _only certain state fields_, watch for changes in a special property (e.g. `props.userID`).

```suggestion
* Alternative 1 : réinitialiser _certains champs de l'état_ en observant les changements d'une prop donnée (i.e. `props.userID`).
```

> * Alternative 2: Vous pouvez également envisager d'utiliser une méthode d'instance impérative via une référence.

Reviewer: * Alternative 2: You can also consider fall back to an imperative instance method using refs.

```suggestion
* Alternative 2 : vous pouvez également envisager d'utiliser une méthode d'instance impérative via une ref.
```

## Et la mémoïsation dans tout ça ? {#what-about-memoization}

> Nous avons également vu que les états dérivés étaient utilisés pour s'assurer qu'une valeur couteuse utilisée dans `render` n'était recalculée que lorsque les entrées changeaient. Cette technique est connus sous le nom de [mémoïsation](https://fr.wikipedia.org/wiki/M%C3%A9mo%C3%AFsation).

Reviewer: ```suggestion
Nous avons également vu des états dérivés utilisés pour s'assurer qu'une valeur coûteuse utilisée dans `render` n'était recalculée que lorsque ses entrées changeaient. Cette technique est connue sous le nom de [mémoïsation](https://fr.wikipedia.org/wiki/M%C3%A9mo%C3%AFsation).
```

> Utiliser les états dérivés pour la mémoïsation n'est pas nécessairement une mauvaise chose mais ce n'est généralement pas la meilleur solution. Il y a une complexité inhérente à la gestion des états dérivés et cette complexité s'accroit avec chaque nouvelle propriété. Par example, si nous ajoutons un deuxième champs dérivé à l'état de notre composant alors nous aurions à gérer séparément les changements de chacun d'eux.

Reviewer: ```suggestion
Utiliser un état dérivé pour la mémoïsation n'est pas nécessairement une mauvaise chose mais ce n'est généralement pas la meilleure solution. Il y a une complexité inhérente à la gestion d'un état dérivé et cette complexité s'accroît avec chaque nouvelle propriété. Par exemple, si nous ajoutions un deuxième champ dérivé à l'état de notre composant alors nous aurions à gérer séparément les changements de chacun d'eux.
```

> Prenons l'example d'un composant qui prend une prop (une liste d'elements) et effectue le rendu des éléments qui correspondent au résultat d'une requête de recherche saisie par l'utilisateur. Nous pourrions utiliser un état dérivé pour stocker la liste filtrée :

Reviewer: ```suggestion
Prenons l'exemple d'un composant qui prend une prop (une liste d'éléments) et affiche les éléments qui correspondent au résultat d'une requête de recherche saisie par l'utilisateur. Nous pourrions utiliser un état dérivé pour stocker la liste filtrée :
```

```js
class Example extends Component {
  state = {
    filterText: "",
  };

>   // *****************************************************************

Reviewer: ```suggestion
  // ******************************************************************
```

>   // REMARQUE: Cet example n'est pas la méthode que nous recommandons.

Reviewer: ```suggestion
  // REMARQUE : cet exemple n'est PAS la méthode que nous recommandons.
```

>   // Voyez plutôt l'example ci-après pour avoir notre recommandation.

Reviewer: ```suggestion
  // Voyez plutôt l'exemple ci-après pour avoir notre recommandation.
```

>   // *****************************************************************

Reviewer: ```suggestion
  // ******************************************************************
```

  static getDerivedStateFromProps(props, state) {

>     // Ré-exécute le filtre à chaque fois que la liste ou le text du filtre change

Reviewer: ```suggestion
    // Ré-exécute le filtre à chaque fois que la liste ou le texte du filtre change.
```

}
```

> Cette solution évite de recalculer `filteredList` plus que nécessaire. Cependant c'est plus compliqué que ça ne devrais car il faut suivre et vérifier les changements individuels de chacunes des props et de l'état pour pouvoir mettre à jour la liste correctement. Dans cet example, nous pouvons simplifier les choses en utilisant `PureComponent` et en déplaçant les operation de filtrage dans la méthode de rendu :

Reviewer: ```suggestion
Cette solution évite de recalculer `filteredList` plus que nécessaire. Elle est toutefois inutilement compliquée car il faut suivre et vérifier les changements individuels de chacune des props et variables d'état pour pouvoir mettre à jour la liste correctement. Dans cet exemple, nous pouvons simplifier les choses en utilisant `PureComponent` et en déplaçant le filtrage dans la méthode de rendu :
```

```js

> // Les PureComponents ne sont rendu que si au moins une props ou une valeur d'état change.

Reviewer: ```suggestion
// Les PureComponents ne se rafraîchissent que si au moins une valeur de prop ou d’état change.
```

// Les changements sont déterminés en faisant une comparaison de surface des clés des objets props et state
class Example extends PureComponent {

>   // L'état n'a besoin de retenir que le text du filtre courrant :

Reviewer: ```suggestion
  // L’état n’a besoin de retenir que le texte actuel du filtre :
```

  };

  render() {

>     // La méthode render de ce PureComponent est appelé seulement si

Reviewer: ```suggestion
    // La méthode render de ce PureComponent est appelée seulement si
```

>     // props.list ou state.filterText ont changés.

Reviewer: ```suggestion
    // props.list ou state.filterText ont changé.
```

}
```

> La solution ci-avant est bien plus propre et simple que la version avec état dérivé. Parfois, ce ne sera pas suffisant—les filtres peuvent êtres lent à appliquer pour les listes longues et `PureComponent` n'empêchera pas le rendu si une autre prop venait à changer. Pour régler ces problèmes nous pouvons ajouter un utilitaire de mémoïsation pour éviter de filtrer inutilement notre liste plusieurs fois.

Reviewer: ```suggestion
La solution ci-avant est bien plus propre et simple que la version avec état dérivé. Parfois, ce ne sera pas suffisant : les filtres peuvent être lents à appliquer pour les listes longues et `PureComponent` n'empêchera pas le rendu si une autre prop venait à changer. Pour régler ces problèmes nous pouvons ajouter un utilitaire de mémoïsation pour éviter de filtrer inutilement notre liste plusieurs fois.
```

```js
import memoize from "memoize-one";

class Example extends Component {

>   // L'état n'a besoin de retenir que le text du filtre courrant :

Reviewer: ```suggestion
  // L’état n’a besoin de retenir que le texte actuel du filtre :
```

  };

  render() {

>     // Calcule la dernière liste filtré. Si les arguments n'ont pas changés depuis

Reviewer: ```suggestion
    // Calcule la dernière liste filtrée. Si les arguments n’ont pas changé depuis
```

}
```

C'est beaucoup plus simple et c'est aussi performant que la version avec état dérivé !

> Quand vous utilisez la mémoïsation, souvenez vous des quelques contraintes suivantes :

Reviewer: ```suggestion
Quand vous utilisez la mémoïsation, souvenez-vous des quelques contraintes suivantes :
```

1. Dans la plupart des cas, vous voudrez **rattacher la fonction mémoïsée à l'instance du composant**. Ça évite que de multiples instances d'un même composant ne réinitialisent les clés de mémoïsation les unes des autres.

> 1. En général, vous voudrez utiliser un utilitaire de mémoïsation avec une **taille de cache limité** afin d'éviter d'éventuelle fuite de mémoire au fur et à mesure que le temps passe. (Dans l'example précédent, nous avons utilisé `memoize-one` car il ne cache que le couple arguments/résultats le plus récent.)

Reviewer: ```suggestion
1. En général, vous voudrez utiliser un utilitaire de mémoïsation avec une **taille de cache limitée** afin d'éviter d'éventuelles fuites de mémoire au fil du temps. (Dans l'exemple précédent, nous avons utilisé `memoize-one` car il ne cache que le couple arguments/résultat le plus récent.)
```

> 1. Aucune des solutions présentées ici ne fonctionnera si `props.list` et recréée à chaque fois que le composant parent est rendu. Cependant, dans la plus part des cas ce système convient.

Reviewer: ```suggestion
1. Aucune des solutions présentées ici ne fonctionnera si `props.list` et recréée à chaque fois que le composant parent se rafraîchit. Cependant, dans la plupart des cas ce système convient.
```

## En conclusion {#in-closing}

> Dans les applications de la vie réelle, les composants comportent un mélange de comportements controlés et non-controlés. Ce n'est pas un problème ! Si chaque valeur a une source de vérité clairement identifiée vous pouvez éviter les anti-pattern que nous avons évoqués ici.

Reviewer: ```suggestion
Dans de véritables applications en production, les composants comportent souvent un mélange de comportements contrôlés et non-contrôlés. Ce n'est pas un problème ! Si chaque valeur a une source de vérité clairement identifiée vous pouvez éviter les anti-patterns que nous avons évoqués ici.
```

> Ça vaut également la peine de redire que `getDerivedStateFromProps` (et les états dérivés en général) est une fonctionnalité avancée qui devrait être utilisée avec parcimonie du fait de sa complexité. Si vos cas d'utilisation sont différents de ces patterns, n'hésitez pas à les partager avec nous sur [GitHub](https://github.com/reactjs/reactjs.org/issues/new) ou [Twitter](https://twitter.com/reactjs) !

Reviewer: ```suggestion
Ça vaut également la peine de rappeler que `getDerivedStateFromProps` (et les états dérivés en général) est une fonctionnalité avancée qui devrait être utilisée avec parcimonie du fait de sa complexité. Si vos cas d'utilisation sont différents de ces schémas, n'hésitez pas à les partager avec nous sur [GitHub](https://github.com/reactjs/reactjs.org/issues/new) ou [Twitter](https://twitter.com/reactjs) !
```
